pyvista/core/utilities/geometric_sources.py

```diff
@@ -2812,3 +2819,1043 @@ def output(self):
+
+    @direction_vectors.setter
+    def direction_vectors(self, vectors):  # numpydoc ignore=GL08
+        self._direction_vectors = _validation.validate_axes(vectors, name='direction_vectors')
+
+    @property
+    def transformation_matrix(self) -> NumpyArray[float]:
+        """Transformation matrix used to orient the axes in space.
+
+        Returns
+        -------
+        numpy.ndarray
+            4x4 transformation matrix.
+        """
+        scale_matrix = np.eye(4)
+        if self.normalized_mode:
+            # Scale proportional to axis length
+            scale_matrix[:3, :3] = np.diag(self.total_length)
+
+        position_matrix = np.eye(4)
+        position_matrix[:3, 3] = self.position
+
+        rotation_matrix = np.eye(4)
+        rotation_matrix[:3, :3] = self.direction_vectors
+
+        # Scale first, then rotate, then move
+        return position_matrix @ rotation_matrix @ scale_matrix
+
+    def _set_geometry(self, part: _PartEnum, geometry: str | pyvista.DataSet):
+        geometry_name, new_datasets = AxesGeometrySource._make_axes_parts(geometry)
+        datasets = self._shaft_datasets if part == _PartEnum.shaft else self._tip_datasets
+        datasets[_AxisEnum.x].copy_from(new_datasets[_AxisEnum.x])
+        datasets[_AxisEnum.y].copy_from(new_datasets[_AxisEnum.y])
+        datasets[_AxisEnum.z].copy_from(new_datasets[_AxisEnum.z])
+        return geometry_name
+
+    def _reset_shaft_and_tip_geometry(self):
+        shaft_radius, shaft_length = self.shaft_radius, self.shaft_length
+        tip_radius, tip_length = (
+            self.tip_radius,
+            self.tip_length,
+        )
+
+        nested_datasets = [self._shaft_datasets, self._tip_datasets]
+        for part_type in _PartEnum:
+            for axis in _AxisEnum:
+                # Reset geometry
+                part = AxesGeometrySource._normalize_part(nested_datasets[part_type][axis])
+
+                # Offset so axis bounds are [0, 1]
+                part.points[:, axis] += 0.5
+
+                # Scale by length along axis, scale by radius off-axis
+                radius, length = (
+                    (shaft_radius, shaft_length)
+                    if part_type == _PartEnum.shaft
+                    else (tip_radius, tip_length)
+                )
+                diameter = radius * 2
+                scale = [diameter] * 3
+                scale[axis] = length[axis]
+                part.scale(scale, inplace=True)
+
+                if part_type == _PartEnum.tip:
+                    # Move tip to end of shaft
+                    part.points[:, axis] += shaft_length[axis]
+
+                if self.symmetric:
+                    # Flip and append to part
+                    origin = [0, 0, 0]
+                    normal = [0, 0, 0]
+                    normal[axis] = 1
+                    flipped = part.flip_normal(normal=normal, point=origin)
+                    part.append_polydata(flipped, inplace=True)
```

Comment:
I recommend to use [itertools.product](https://docs.python.org/3/library/itertools.html#itertools.product) in this kind of situation.
```suggestion
        for part_type, axis in itertools.product(_PartEnum, _AxisEnum):
            for axis in _AxisEnum:
            # Reset geometry
            part = AxesGeometrySource._normalize_part(nested_datasets[part_type][axis])

            # Offset so axis bounds are [0, 1]
            part.points[:, axis] += 0.5

            # Scale by length along axis, scale by radius off-axis
            radius, length = (
                (shaft_radius, shaft_length)
                if part_type == _PartEnum.shaft
                else (tip_radius, tip_length)
            )
            diameter = radius * 2
            scale = [diameter] * 3
            scale[axis] = length[axis]
            part.scale(scale, inplace=True)

            if part_type == _PartEnum.tip:
                # Move tip to end of shaft
                part.points[:, axis] += shaft_length[axis]

            if self.symmetric:
                # Flip and append to part
                origin = [0, 0, 0]
                normal = [0, 0, 0]
                normal[axis] = 1
                flipped = part.flip_normal(normal=normal, point=origin)
                part.append_polydata(flipped, inplace=True)
```